File: flutter_mort/lib/core/routing/notification_destination.dart
import '../../data/models/profile.dart';

final _notificationUuid = RegExp(
  r'^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$',
  caseSensitive: false,
);

String notificationDestination(Map<String, dynamic> data, UserRole? role) {
  final threadId = _uuid(data['threadId']);
  if (threadId != null) return '/messages/$threadId';

  final supportTicketId = _uuid(data['supportTicketId']);
  if (supportTicketId != null) return '/support/ticket/$supportTicketId';

  if (_uuid(data['reviewId']) != null) return '/settings/reviews';
  if (_uuid(data['guardianLinkId']) != null) {
    return '/settings/guardian-mode';
  }

  final applicationId = _uuid(data['applicationId']);
  if (applicationId != null) {
    return switch (role) {
      UserRole.adult => '/adult/applicants/$applicationId',
      UserRole.guardian => '/guardian/approvals/$applicationId',
      UserRole.teen => '/teen/applications/$applicationId',
      _ => '/account-status',
    };
  }

  final jobId = _uuid(data['jobId']);
  if (jobId != null) {
    return switch (role) {
      UserRole.adult => '/adult/jobs/$jobId',
      UserRole.teen => '/teen/jobs/$jobId',
      _ => '/account-status',
    };
  }

  if (_uuid(data['safetyPingId']) != null) {
    return switch (role) {
      UserRole.admin => '/admin/safety-pings',
      UserRole.guardian => '/guardian/safety-pings',
      UserRole.teen => '/teen/safety',
      _ => '/account-status',
    };
  }
  return '/account-status';
}

String? _uuid(Object? value) {
  final candidate = value?.toString().trim();
  return candidate != null && _notificationUuid.hasMatch(candidate)
      ? candidate.toLowerCase()
      : null;
}

File: swift_mort/MORT/Features/Notifications/NotificationDestination.swift
import Foundation

enum NotificationLinkResolution: Equatable {
    case notNotification
    case destination(AppRoute)
}

enum NotificationDestinationResolver {
    private static let notificationHosts = Set(["notification", "notifications"])

    static func destination(for data: [String: JSONValue], role: UserRole?) -> AppRoute {
        if let id = uuid(data["threadId"]) {
            return .messageThread(id)
        }

        if isPresent(data["supportTicketId"]) {
            return role == .admin ? .adminQueue(.support) : .support
        }

        if isPresent(data["reportId"]) {
            return role == .admin ? .adminQueue(.reports) : .activity
        }

        if isPresent(data["reviewId"]) {
            return role == .admin ? .adminQueue(.reviews) : .reviews
        }

        if isPresent(data["guardianLinkId"]) {
            return role == .teen || role == .guardian ? .guardianMode : .activity
        }

        if let id = uuid(data["applicationId"]) {
            switch role {
            case .some(.teen): return .applicationDetail(id, .teen)
            case .some(.adult): return .applicationDetail(id, .adult)
            case .some(.guardian): return .applicationDetail(id, .guardian)
            case .some(.admin), .none: return .activity
            }
        }

        if isPresent(data["safetyPingId"]) {
            switch role {
            case .some(.admin): return .adminQueue(.safetyPings)
            case .some(.guardian): return .guardianSafetyPings
            case .some(.teen), .some(.adult), .none: return .safetyCenter
            }
        }

        if let id = uuid(data["jobId"]) {
            return .jobDetail(id)
        }

        if isPresent(data["verificationId"]) || isPresent(data["verificationStatus"]) {
            switch role {
            case .some(.admin): return .adminQueue(.verifications)
            case .some(.adult): return .verification
            case .some(.teen), .some(.guardian), .none: return .activity
            }
        }

        return .activity
    }

    static func linkResolution(for url: URL, role: UserRole?) -> NotificationLinkResolution {
        guard url.scheme?.lowercased() == "mort",
              let host = url.host?.lowercased(),
              notificationHosts.contains(host)
        else {
            return .notNotification
        }

        var data: [String: JSONValue] = [:]
        for item in URLComponents(url: url, resolvingAgainstBaseURL: false)?.queryItems ?? [] {
            guard let value = item.value, !value.isEmpty else { continue }
            data[item.name] = .string(value)
        }
        return .destination(destination(for: data, role: role))
    }

    private static func uuid(_ value: JSONValue?) -> UUID? {
        guard let text = value?.stringValue else { return nil }
        return UUID(uuidString: text)
    }

    private static func isPresent(_ value: JSONValue?) -> Bool {
        guard let value else { return false }
        switch value {
        case .null: return false
        case let .string(text): return !text.isEmpty
        default: return true
        }
    }
}
